### PR TITLE
Add array_slice function

### PR DESCRIPTION
**Pull request by jacobgao**

### CODE REVIEW

**Comment by haozhun** [comment]
Change the commit message and title of this pull request to "Add array_slice function".

* Always start commit message with capital letter
* The concept of "User-defined" function doesn't really exist in Presto

**Comment by haozhun** [comment]
If I remember correctly, we decided to throw an exception when `length < 0` and return an empty array in all other failure cases.

**Comment by haozhun** [comment]
Do something similar to this (taken from `DateTimeFunctions`)

    checkCondition((zoneOffset % 60_000) == 0, INVALID_FUNCTION_ARGUMENT, "Invalid time zone offset interval: interval contains seconds");

This throws a `PrestoException` with `INVALID_FUNCTION_ARGUMENT` when condition is not met.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraySliceFunction.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+import static com.facebook.presto.type.TypeUtils.readStructuralBlock;
+import static com.facebook.presto.util.Failures.checkCondition;
+import static com.facebook.presto.util.Reflection.methodHandle;
+
+public final class ArraySliceFunction
+        extends ParametricScalar
+{
+    public static final ArraySliceFunction ARRAY_SLICE_FUNCTION = new ArraySliceFunction();
+    private static final String FUNCTION_NAME = "slice";
+    private static final Signature SIGNATURE = new Signature(FUNCTION_NAME, ImmutableList.of(comparableTypeParameter("E")), "array<E>", ImmutableList.of("array<E>", "bigint", "bigint"), false, false);
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraySliceFunction.java]: this should be `typeParameter("E")` otherwise it will only work for arrays of orderable types

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraySliceFunction.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+    @Override
+    public FunctionInfo specialize(Map<String, Type> types, int arity, TypeManager typeManager, FunctionRegistry functionRegistry)
+    {
+        Type type = types.get("E");
+        MethodHandle methodHandle = METHOD_HANDLE.bindTo(type);
+        Signature signature = new Signature(FUNCTION_NAME,
+                parameterizedTypeName("array", type.getTypeSignature()),
+                parameterizedTypeName("array", type.getTypeSignature()), parseTypeSignature("bigint"), parseTypeSignature("bigint"));
+        return new FunctionInfo(signature, getDescription(), isHidden(), methodHandle, isDeterministic(), true, ImmutableList.of(false, false, false));
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraySliceFunction.java]: the second to last argument should be `false` since this cannot produce a `NULL` result unless one of the parameters is null

**Comment by cberner** [comment]
This needs documentation (add it to `array.rst`)

**Comment by cberner** [comment]
Arrays are 1-indexed, so the offset needs to be a 1-indexed argument. It should also throw an exception if the offset is zero

File: presto-main/src/test/java/com/facebook/presto/type/TestArrayOperators.java [hunk 3]
```diff
@@ -373,6 +373,20 @@ public void testSort()
...
+    {
+        assertFunction("SLICE(ARRAY [1, 2, 3, 4, 5], 0, 4)", new ArrayType(BIGINT), ImmutableList.of(1L, 2L, 3L, 4L));
+        assertFunction("SLICE(ARRAY [1, 2], 0, 4)", new ArrayType(BIGINT), ImmutableList.of(1L, 2L));
+        assertFunction("SLICE(ARRAY [1, 2, 3, 4, 5], 2, 2)", new ArrayType(BIGINT), ImmutableList.of(3L, 4L));
+        assertFunction("SLICE(ARRAY ['1', '2', '3', '4'], 1, 1)", new ArrayType(VARCHAR), ImmutableList.of("2"));
+        assertFunction("SLICE(ARRAY [1, 2, 3, 4], 2, 3)", new ArrayType(BIGINT), ImmutableList.of(3L, 4L));
+        assertFunction("SLICE(ARRAY [1, 2, 3, 4], -3, 3)", new ArrayType(BIGINT), ImmutableList.of(2L, 3L, 4L));
+        assertFunction("SLICE(ARRAY [1, 2, 3, 4], -3, 5)", new ArrayType(BIGINT), ImmutableList.of(2L, 3L, 4L));
+        assertFunction("SLICE(ARRAY [1, 2, 3, 4], 1, 0)", new ArrayType(BIGINT), ImmutableList.of());
+        assertFunction("SLICE(ARRAY [1, 2, 3, 4], -2, 0)", new ArrayType(BIGINT), ImmutableList.of());
```
Comment (cberner) [inline presto-main/src/test/java/com/facebook/presto/type/TestArrayOperators.java]: Add a test that this fails for a offset of zero

**Comment by cberner** [comment]
Squashed the commits and merged :)
